source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,594 @@
+        sizeof(DeviceInfo->Alignment), &DeviceInfo->Alignment, nullptr));
+
+    // Allocate shadow memory
+    UR_CALL(allocShadowMemory(Context, DeviceInfo));
+
+    auto ContextInfo = getContextInfo(Context);
+    std::scoped_lock<ur_shared_mutex> Guard(ContextInfo->Mutex);
+    ContextInfo->DeviceMap.emplace(Device, std::move(DeviceInfo));
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::addQueue(ur_context_handle_t Context,
+                                           ur_queue_handle_t Queue) {
+    auto QueueInfo = std::make_shared<ur_sanitizer_layer::QueueInfo>();
+    QueueInfo->LastEvent = nullptr;
+
+    auto ContextInfo = getContextInfo(Context);
+    std::scoped_lock<ur_shared_mutex> Guard(ContextInfo->Mutex);
+    ContextInfo->QueueMap.emplace(Queue, std::move(QueueInfo));
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::removeQueue(ur_context_handle_t Context,
+                                              ur_queue_handle_t Queue) {
+    auto ContextInfo = getContextInfo(Context);
+    std::scoped_lock<ur_shared_mutex> Guard(ContextInfo->Mutex);
+    assert(ContextInfo->QueueMap.find(Queue) != ContextInfo->QueueMap.end());
+    ContextInfo->QueueMap.erase(Queue);
+    return UR_RESULT_SUCCESS;
+}
+
+void SanitizerInterceptor::prepareLaunch(ur_queue_handle_t Queue,
+                                         ur_kernel_handle_t Kernel) {
+    auto Context = getContext(Queue);
+    auto Device = getDevice(Queue);
+    auto Program = getProgram(Kernel);
+
+    auto ContextInfo = getContextInfo(Context);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::scoped_lock<ur_mutex> Guard(QueueInfo->Mutex);
+    ur_event_handle_t LastEvent = QueueInfo->LastEvent;
+
+    do {
+        // Set global variable to program
+        auto EnqueueWriteGlobal = [&](const char *Name, const void *Value) {
+            ur_event_handle_t NewEvent{};
+            uint32_t NumEvents = LastEvent ? 1 : 0;
+            const ur_event_handle_t *EventsList =
+                LastEvent ? &LastEvent : nullptr;
+            auto Result =
+                context.urDdiTable.Enqueue.pfnDeviceGlobalVariableWrite(
+                    Queue, Program, Name, false, sizeof(uptr), 0, Value,
+                    NumEvents, EventsList, &NewEvent);
+            if (Result != UR_RESULT_SUCCESS) {
+                context.logger.warning("Device Global[{}] Write Failed: {}",
+                                       Name, Result);
+                return false;
+            }
+            LastEvent = NewEvent;
+            return true;
+        };
+
+        // Device shadow memory offset
+        EnqueueWriteGlobal(kSPIR_AsanShadowMemoryGlobalStart,
+                           &DeviceInfo->ShadowOffset);
+        EnqueueWriteGlobal(kSPIR_AsanShadowMemoryGlobalEnd,
+                           &DeviceInfo->ShadowOffsetEnd);
+    } while (false);
```

Comment:
why the single iteration loop?

Reply:
it's not meaningful now. I'll remove this loop.